api/basic/get-pins-router.ts

```diff
@@ -24,68 +25,84 @@ export async function handler (event: APIGatewayProxyEventV2, context: Context):
+
+  /* eslint-disable @typescript-eslint/prefer-nullish-coalescing */
+  /* eslint-disable @typescript-eslint/strict-boolean-expressions */
+  let validationError: ValidationError | undefined = validateRoutingConfiguration({ legacyClusterIpfsUrl, pickupUrl })
 
   if (validationError != null) {
-    return { statusCode: validationError.statusCode, body: JSON.stringify(validationError.body) }
+    logger.error({ err: validationError, code: validationError.code }, 'Validation config error on get pin router')
+    return toResponseError(500, 'INTERNAL_SERVER_ERROR')
   }
 
-  logger.trace('Parameters are valid')
+  validationError = validateGetPinsParameters({ cids })
 
+  if (validationError) {
+    logger.error({ err: validationError, code: validationError.code }, 'Validation event params error on get pin router')
+    return toResponseError(400, 'BAD_REQUEST', validationError.message)
+  }
+
+  let cidsNotFound: any[] = []
+  let foundInPickup: Record<string, ClusterGetResponseBody> = {}
   try {
     logger.trace('Get from pickup')
     const pickupResponse = await fetchGetPins({ cids, endpoint: pickupUrl, isInternal: true, token })
 
     logger.trace(pickupResponse, 'Pickup response')
 
-    const foundInPickup = pickupResponse.body?.filter(item => (Object.values(item.peer_map).filter(pin => pin.status !== 'unpinned').length > 0)).reduce((acc: Record<string, ClusterStatusResponse>, next) => {
-      acc[next.cid] = next
-      return acc
-    }, {})
+    if (pickupResponse) {
+      foundInPickup = pickupResponse.filter(item => (Object.values(item.peer_map).filter(pin => pin.status !== 'unpinned').length > 0)).reduce((acc: Record<string, ClusterGetResponseBody>, next) => {
+        acc[next.cid] = next
+        return acc
+      }, {})
 
-    logger.trace(foundInPickup, 'Found in pickup')
+      logger.trace(foundInPickup, 'Found in pickup')
 
-    const cidNotFound = pickupResponse.body?.filter(item =>
-      (Object.values(item.peer_map)
-        .filter(pin => pin.status !== 'unpinned').length === 0)
-    ).map(item => item.cid)
+      cidsNotFound = pickupResponse.filter(item =>
+        (Object.values(item.peer_map)
+          .filter(pin => pin.status !== 'unpinned').length === 0)
+      ).map(item => item.cid)
 
-    logger.trace(cidNotFound, 'Cid not found in pickup')
+      logger.trace(cidsNotFound, 'Cid not found in pickup')
 
-    if (!cidNotFound.length) {
-      return {
-        statusCode: 200,
-        body: pickupResponse.body.map(pin => JSON.stringify(pin)).join('\n')
+      if (!cidsNotFound.length) {
+        logger.info({ code: 'FROM_PICKUP', cids }, 'Get pins from pickup')
+        return toResponseFromString(pickupResponse.map(pin => JSON.stringify(pin)).join('\n'))
       }
     }
+  } catch (err: any) {
+    logger.error({ err, code: 'FROM_PICKUP' }, 'Error on get pins router - pickup')
+  }
+
+  try {
+    const missingCids = [...cidsNotFound].concat(cids.split(',').filter(cid => cidsNotFound.includes(cid)))
```

Comment:
@codeflyer note is requesting the missing pins from pickup

Reply:
Pickup is alway returning the full list

Reply:
This can be an additional check, but it's not required.